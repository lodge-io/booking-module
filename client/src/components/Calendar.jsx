
import moment from 'moment';

const { React } = window;

const utcMoment = moment.utc;

const SHORT_DAY_NAMES = ['Su', 'Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa'];
const MONTH_NAMES = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'];

function getCurrMonth() {
  return new Date(Date.now()).getMonth();
}

function getCurrYear() {
  return new Date(Date.now()).getFullYear();
}

function getLastMonthDate(month, year) {
  return (new Date(year, month + 1, 0)).getDate();
}

function numToMonth(num) {
  return MONTH_NAMES[num];
}

function getMonthStartWeekday(month, year) {
  return new Date(year, month).getDay();
}

function createDate(year, month, day) {
  const strYear = `${year}`;
  const strMonth = month >= 10 ? `${month + 1}` : `0${month + 1}`;
  const strDay = day >= 10 ? `${day}` : `0${day}`;
  return utcMoment(`${strYear}-${strMonth}-${strDay}`);
}

const CalendarBox = styled.div`
  padding: 0.25em 0.25em;
  text-align: center;
`;
const TableD = styled.td`
  border: 1px solid #EBEBEB;
  width: 42px; height:42px;
  background-color: white;  
  &:hover{
    background-color: #00A699;
  }
`;
TableD.displayName = 'TableD';

const TableDDisabled = styled.td`
  border: 1px solid #EBEBEB;
  width: 42px; height:42px;
  color: #D8D8D8;
`;
TableDDisabled.displayName = 'TableDDisabled';

const ArrowSpan = styled.span`
  flex-grow: 1;
  text-align: center;
  border: 1px solid black;
  border-radius: 3px;
  padding-top: 3px;
  padding-bottom: 3px;
`;
const MonthSpan = styled.span`
  flex-grow: 8;
  text-align: center;
`;


const Head = styled.th`
  width: 14%;
`;
const Table = styled.table`
  width:100%;
  padding:0px;
  border-collapse: collapse;
`;
const TopRow = styled.span`
  display: flex;
  flex-direction: row;
`;
const TableHolder = styled.div`
  width: 300px;
  padding: 20px;
  border-radius: 3px;
  user-select: none;
`;

const CalendarDate = ({ year, month, date, available, inputDate, selecting }) => {
  if (available) {
    return (
      <TableD key={date}>
        <CalendarBox
          key={date}
          className="clickableDate"
          onClick={() => {
            inputDate(createDate(year, month, date));
          }
          }
        >
          {date}
        </CalendarBox>
      </TableD>
    );
  }

  return (
    <TableDDisabled key={date}>
      <CalendarBox
        key={date}
      >
        {date}
      </CalendarBox>
    </TableDDisabled>
  );

};

class Calendar extends React.Component {
  constructor(props) {
    super(props);
    this.state = { month: getCurrMonth(), year: getCurrYear() };
    this.isValidDate = this.isValidDate.bind(this);
  }

  nextMonth() {
    this.setState((prevState) => {
      let { month, year } = prevState;
      if (month === 11) {
        month = 0;
        year += 1;
      } else {
        month += 1;
      }
      return { month, year };
    });
  }

  lastMonth() {
    this.setState((prevState) => {
      let { month, year } = prevState;
      if (month === 0) {
        month = 11;
        year -= 1;
      } else {
        month -= 1;
      }
      return { month, year };
    });
  }

  isValidDate(year, month, day) {
    const { bookings, selecting } = this.props;
    const date = utcMoment(`${year}/${month + 1}/${day}`);
    const now = utcMoment().startOf('day');
    if (date - now < 0) {
      return false;
    }
    if (bookings.length === 0) {
      return true;
    }
    let i = 0;
    if (selecting === 0) {
      while (i < bookings.length && bookings[i].endDate.diff(date) <= 0) { i += 1; }
      if (i === bookings.length) { return true; }
      if (bookings[i].startDate <= date) { return false; }
    } else {
      while (i < bookings.length && bookings[i].endDate.diff(date) < 0) { i += 1; }
      if (i === bookings.length) { return true; }
      if (bookings[i].startDate < date) { return false; }
    }
    return true;
  }

  render() {
    const { year, month } = this.state;
    const { inputDate } = this.props;
    const monthName = numToMonth(month);
    let date = 1;
    const lastDay = getLastMonthDate(month, year);
    const firstDay = getMonthStartWeekday(month, year);
    const table = [];
    let emptyNum = -1;

    while (date <= lastDay) {
      const week = [];
      for (let i = 0; i < 7; i += 1) {
        if ((date === 1 && i < firstDay) || date > lastDay) {
          week.push(<td key={emptyNum} />);
          emptyNum -= 1;
        } else {
          const myDate = date;
          week.push(
            <CalendarDate
              available={(this.isValidDate(year, month, myDate))}
              year={year}
              month={month}
              date={myDate}
              inputDate={inputDate}
            />,
          );
          date += 1;
        }
      }
      table.push(week);
    }
    return (
      <TableHolder>
        <TopRow>
          <ArrowSpan className="lastMonth" onClick={() => this.lastMonth()}>&lt;</ArrowSpan>
          <MonthSpan className="calTitleSpan">
            {monthName}
            {' '}
            {year}
          </MonthSpan>
          <ArrowSpan className="nextMonth" onClick={() => this.nextMonth()}>&gt;</ArrowSpan>
        </TopRow>
        <Table>
          <tbody>
            <tr key="header">
              {SHORT_DAY_NAMES.map(val => <Head key={val}>{val}</Head>)}
            </tr>
            {table.map((week, i) => <tr key={`weekRow:${i}`}>{week.map(val => val)}</tr>)}
          </tbody>
        </Table>
      </TableHolder>
    );
  }
}

export default Calendar;
