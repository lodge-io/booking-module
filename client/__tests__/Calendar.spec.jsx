import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import moment from 'moment';
import Calendar from '../src/components/Calendar';

describe('Calendar', () => {
  it('renders in the dom', () => {
    const res = shallow(<Calendar />).find('.calTitleSpan').contains(2019);
    expect(res).toBeTruthy();
  });

  it('should log when a date is hovered, clicked or focused', () => {

    const mockListener = jest.fn(() => {});

    const wrapper = shallow(<Calendar inputDate={mockListener}/>);
    const dates = wrapper.find('.a-date');
    dates.at(0).simulate('click');
    dates.at(5).simulate('click');
    dates.at(23).simulate('click');
    dates.at(11).simulate('click');

    expect(mockListener.mock.calls.length).toBe(4);
    expect(mockListener.mock.calls[0][0] instanceof moment).toBeTruthy();
    expect(mockListener.mock.calls[1][0] instanceof moment).toBeTruthy();
    expect(mockListener.mock.calls[2][0] instanceof moment).toBeTruthy();
    expect(mockListener.mock.calls[3][0] instanceof moment).toBeTruthy();
  });

  it('should change month when next/last button clicked', () => {
    const wrapper = shallow(<Calendar />);
    let testYear = new Date(Date.now()).getFullYear();
    let testMonth = moment().format('MMMM');
    let title = wrapper.find('.calTitleSpan');
    expect(title.contains(testYear)).toBeTruthy();
    expect(title.contains(testMonth)).toBeTruthy();

    const last = wrapper.find('.lastMonth');
    const next = wrapper.find('.nextMonth');
    last.simulate('click');
    title = wrapper.find('.calTitleSpan');
    testMonth = moment().subtract(1, 'months').format('MMMM');
    expect(title.contains(testMonth)).toBeTruthy();

    for (let i = 0; i < 11; i += 1) {
      last.simulate('click');
    }
    title = wrapper.find('.calTitleSpan');
    testYear = parseInt(moment().subtract(1, 'years').format('YYYY'), 10);

    expect(title.contains(testYear)).toBeTruthy();

    next.simulate('click');
    title = wrapper.find('.calTitleSpan');
    testMonth = moment().add(1, 'months').format('MMMM');
    expect(title.contains(testMonth)).toBeTruthy();
    expect(title.contains(testYear)).toBeTruthy();
    for (let i = 0; i < 11; i += 1) {
      next.simulate('click');
    }

    title = wrapper.find('.calTitleSpan');
    testMonth = moment().format('MMMM');
    testYear = parseInt(moment().format('YYYY'), 10);
    expect(title.contains(testMonth)).toBeTruthy();
    expect(title.contains(testYear)).toBeTruthy();  
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(<Calendar />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
