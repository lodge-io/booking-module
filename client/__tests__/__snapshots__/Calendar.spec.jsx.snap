// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar renders correctly 1`] = `
<div
  className="sc-htoDjs ixOviw"
>
  <span
    className="sc-gzVnrw jKdxIN"
  >
    <span
      className="lastMonth sc-bxivhb dwoOdW"
      onClick={[Function]}
    >
      &lt;
    </span>
    <span
      className="calTitleSpan sc-ifAKCX djMrCb"
    >
      April
       
      2019
    </span>
    <span
      className="nextMonth sc-bxivhb dwoOdW"
      onClick={[Function]}
    >
      &gt;
    </span>
  </span>
  <table
    className="sc-bZQynM fOTTne"
  >
    <tbody>
      <tr>
        <th
          className="sc-EHOje itbzhn"
        >
          Su
        </th>
        <th
          className="sc-EHOje itbzhn"
        >
          Mo
        </th>
        <th
          className="sc-EHOje itbzhn"
        >
          Tu
        </th>
        <th
          className="sc-EHOje itbzhn"
        >
          We
        </th>
        <th
          className="sc-EHOje itbzhn"
        >
          Th
        </th>
        <th
          className="sc-EHOje itbzhn"
        >
          Fr
        </th>
        <th
          className="sc-EHOje itbzhn"
        >
          Sa
        </th>
      </tr>
      <tr>
        <td />
        <td
          className="sc-htpNat llddHq"
        >
          <div
            className="sc-bdVaJa faROED"
          >
            1
          </div>
        </td>
        <td
          className="sc-htpNat llddHq"
        >
          <div
            className="sc-bdVaJa faROED"
          >
            2
          </div>
        </td>
        <td
          className="sc-htpNat llddHq"
        >
          <div
            className="sc-bdVaJa faROED"
          >
            3
          </div>
        </td>
        <td
          className="sc-htpNat llddHq"
        >
          <div
            className="sc-bdVaJa faROED"
          >
            4
          </div>
        </td>
        <td
          className="sc-htpNat llddHq"
        >
          <div
            className="sc-bdVaJa faROED"
          >
            5
          </div>
        </td>
        <td
          className="sc-htpNat llddHq"
        >
          <div
            className="sc-bdVaJa faROED"
          >
            6
          </div>
        </td>
      </tr>
      <tr>
        <td
          className="sc-bwzfXH ivOUJN"
        >
          <div
            className="clickableDate sc-bdVaJa faROED"
            onClick={[Function]}
          >
            7
          </div>
        </td>
        <td
          className="sc-bwzfXH ivOUJN"
        >
          <div
            className="clickableDate sc-bdVaJa faROED"
            onClick={[Function]}
          >
            8
          </div>
        </td>
        <td
          className="sc-bwzfXH ivOUJN"
        >
          <div
            className="clickableDate sc-bdVaJa faROED"
            onClick={[Function]}
          >
            9
          </div>
        </td>
        <td
          className="sc-bwzfXH ivOUJN"
        >
          <div
            className="clickableDate sc-bdVaJa faROED"
            onClick={[Function]}
          >
            10
          </div>
        </td>
        <td
          className="sc-bwzfXH ivOUJN"
        >
          <div
            className="clickableDate sc-bdVaJa faROED"
            onClick={[Function]}
          >
            11
          </div>
        </td>
        <td
          className="sc-bwzfXH ivOUJN"
        >
          <div
            className="clickableDate sc-bdVaJa faROED"
            onClick={[Function]}
          >
            12
          </div>
        </td>
        <td
          className="sc-bwzfXH ivOUJN"
        >
          <div
            className="clickableDate sc-bdVaJa faROED"
            onClick={[Function]}
          >
            13
          </div>
        </td>
      </tr>
      <tr>
        <td
          className="sc-bwzfXH ivOUJN"
        >
          <div
            className="clickableDate sc-bdVaJa faROED"
            onClick={[Function]}
          >
            14
          </div>
        </td>
        <td
          className="sc-bwzfXH ivOUJN"
        >
          <div
            className="clickableDate sc-bdVaJa faROED"
            onClick={[Function]}
          >
            15
          </div>
        </td>
        <td
          className="sc-bwzfXH ivOUJN"
        >
          <div
            className="clickableDate sc-bdVaJa faROED"
            onClick={[Function]}
          >
            16
          </div>
        </td>
        <td
          className="sc-bwzfXH ivOUJN"
        >
          <div
            className="clickableDate sc-bdVaJa faROED"
            onClick={[Function]}
          >
            17
          </div>
        </td>
        <td
          className="sc-bwzfXH ivOUJN"
        >
          <div
            className="clickableDate sc-bdVaJa faROED"
            onClick={[Function]}
          >
            18
          </div>
        </td>
        <td
          className="sc-bwzfXH ivOUJN"
        >
          <div
            className="clickableDate sc-bdVaJa faROED"
            onClick={[Function]}
          >
            19
          </div>
        </td>
        <td
          className="sc-bwzfXH ivOUJN"
        >
          <div
            className="clickableDate sc-bdVaJa faROED"
            onClick={[Function]}
          >
            20
          </div>
        </td>
      </tr>
      <tr>
        <td
          className="sc-bwzfXH ivOUJN"
        >
          <div
            className="clickableDate sc-bdVaJa faROED"
            onClick={[Function]}
          >
            21
          </div>
        </td>
        <td
          className="sc-bwzfXH ivOUJN"
        >
          <div
            className="clickableDate sc-bdVaJa faROED"
            onClick={[Function]}
          >
            22
          </div>
        </td>
        <td
          className="sc-bwzfXH ivOUJN"
        >
          <div
            className="clickableDate sc-bdVaJa faROED"
            onClick={[Function]}
          >
            23
          </div>
        </td>
        <td
          className="sc-bwzfXH ivOUJN"
        >
          <div
            className="clickableDate sc-bdVaJa faROED"
            onClick={[Function]}
          >
            24
          </div>
        </td>
        <td
          className="sc-bwzfXH ivOUJN"
        >
          <div
            className="clickableDate sc-bdVaJa faROED"
            onClick={[Function]}
          >
            25
          </div>
        </td>
        <td
          className="sc-bwzfXH ivOUJN"
        >
          <div
            className="clickableDate sc-bdVaJa faROED"
            onClick={[Function]}
          >
            26
          </div>
        </td>
        <td
          className="sc-bwzfXH ivOUJN"
        >
          <div
            className="clickableDate sc-bdVaJa faROED"
            onClick={[Function]}
          >
            27
          </div>
        </td>
      </tr>
      <tr>
        <td
          className="sc-bwzfXH ivOUJN"
        >
          <div
            className="clickableDate sc-bdVaJa faROED"
            onClick={[Function]}
          >
            28
          </div>
        </td>
        <td
          className="sc-bwzfXH ivOUJN"
        >
          <div
            className="clickableDate sc-bdVaJa faROED"
            onClick={[Function]}
          >
            29
          </div>
        </td>
        <td
          className="sc-bwzfXH ivOUJN"
        >
          <div
            className="clickableDate sc-bdVaJa faROED"
            onClick={[Function]}
          >
            30
          </div>
        </td>
        <td />
        <td />
        <td />
        <td />
      </tr>
    </tbody>
  </table>
</div>
`;
